docs/src/literate-tutorials/gpu_qp_heat_equation_v2.jl

```diff
@@ -0,0 +1,274 @@
+    (;f, K, assembler, Ke, fe) = create_buffers(cellvalues,dh)
+    # Loop over all cels
+    for cell in CellIterator(dh)
+        fill!(Ke, 0)
+        fill!(fe, 0)
+        if QPiter
+            #reinit!(cellvalues, getcoordinates(cell))
+            assemble_element_qpiter!(Ke, fe, cellvalues,getcoordinates(cell))
+        else
+            # Reinitialize cellvalues for this cell
+            reinit!(cellvalues, cell)
+            # Compute element contribution
+            assemble_element_std!(Ke, fe, cellvalues)
+        end
+        # Assemble Ke and fe into K and f
+        assemble!(assembler, celldofs(cell), Ke, fe)
+    end
+    return K, f
+end
+
+
+"""
+    get_local_sides(e_color,n_basefuncs)
+
+Get the local stiffness matrix and force vector for the element `e_color` in the grid.
+"""
+@inline function get_local_sides(ke_colors, fe_colors,e_color,n_basefuncs)
+    start_row_index = (e_color-1)*n_basefuncs + 1
+    end_row_index = start_row_index+(n_basefuncs-1)
+    start_row_index, end_row_index
+    return @view(ke_colors[start_row_index:end_row_index,1:n_basefuncs]), @view(fe_colors[start_row_index:end_row_index])
+end
+
+
+#=NVTX.@annotate=# function assemble_element_gpu!(ke_colors,fe_colors,assembler,cv,dh,n_cells_colored, eles_colored)
+    tx = threadIdx().x
+    bx = blockIdx().x
+    bd = blockDim().x
+    # e_color is the global index of the thread in the GPU scheme.
+    e_color = tx + (bx-Int32(1))*bd # element number per color
+
+    e_color ≤ n_cells_colored || return nothing
+    n_basefuncs = getnbasefunctions(cv)
+    # e is the global index of the finite element in the grid.
+    e = eles_colored[e_color]
+    cell_coords = getcoordinates(dh.grid, e)
+
+    ke,fe = get_local_sides(ke_colors, fe_colors,e_color,n_basefuncs)
+    fill!(ke, 0.0f0)
+    fill!(fe, 0.0f0)
+     #Loop over quadrature points
+     for qv in Ferrite.QuadratureValuesIterator(cv,cell_coords)
+        ## Get the quadrature weight
+        dΩ = getdetJdV(qv)
+        ## Loop over test shape functions
+        for i in 1:n_basefuncs
+            δu  = shape_value(qv, i)
+            ∇δu = shape_gradient(qv, i)
+            ## Add contribution to fe
+            fe[i] += δu * dΩ
+            ## Loop over trial shape functions
+            for j in 1:n_basefuncs
+                ∇u = shape_gradient(qv, j)
+                ## Add contribution to Ke
+                ke[i,j] += (∇δu ⋅ ∇u) * dΩ
+            end
+        end
+    end
+
+    ## Assemble Ke into Kgpu ##
+    assemble!(assembler, celldofs(dh,e),ke,fe)
+
+    return nothing
+end
+
+
+function allocate_local_matrices(colors,cv)
+    max_color_size = colors .|> length |> maximum
+    n_basefuncs = getnbasefunctions(cv)
+    cols = n_basefuncs
+    rows = n_basefuncs * max_color_size
+    # allocate maximum possible memory to incorporate all local matrices for each color.
+    # It will be mutated in each color.
+    ke = CUDA.zeros(Float32, rows, cols)
+    fe = CUDA.zeros(Float32, rows)
```

Comment:
Is there any advantage over using the following
```suggestion
    ke = CUDA.zeros(Float32, n_elements, n_basefuncs, n_basefuncs)
    fe = CUDA.zeros(Float32, n_elements, n_basefuncs)
```
?

My idea was to swap this 3D array then later with a custom array type which can have varying inner dimensions.

Reply:
You told me about this change last time, but I completely overlooked it, no 3D would be easier and more descriptive. Thanks 🫡